### DISCUSSION TITLE
Working script returning errors since last windows upgrade

### DISCUSSION THREAD

**Discussion by Montreal666** [ImageMagick/ImageMagick#6418] [Help]
### ImageMagick version

7.1.1-11

### Operating system, version and so on

Windows 11 - up to date

### Description

I have the following script which had been working flawlessly for multiple months;

setlocal enabledelayedexpansion

for %%F in (*.jpg) do (

  for /F "usebackq" %%L in (`magick -ping "%%F" ^
    -format "DoIt=%%[fx:%w==1001&&%h==1001?0:1]" ^
    info:`) do set %%L

  if !DoIt!==1 (
    magick "%%F" -fuzz 1%% -trim +repage -resize 1001x1001^^ -gravity Center -crop 1001x1001+0+0 -format jpg -strip -gaussian-blur 0.05 -quality 80 +repage "%%F"
  )

)

following a recent windows upgrade, I now get the following errors for each item processed:

D:\Dropbox\ALBUM ART>(
for /F "usebackq" %L in (`magick -ping "Papa Roach - Last Resort.jpg"   -format "DoIt=%[fx:h==1001?0:1]"   info:`) do set %L
 if !DoIt! == 1 (magick "Papa Roach - Last Resort.jpg" -fuzz 1% -trim +repage -resize 1001x1001^ -gravity Center -crop 1001x1001+0+0 -format jpg -strip -gaussian-blur 0.05 -quality 80 +repage "Papa Roach - Last Resort.jpg" )
)
**magick: GetHslInt failure 0 0,0 @ error/fx.c/ExecuteRPN/3151.
magick: ExecuteRPN failed   @ error/fx.c/FxEvaluateChannelExpression/4002.**

I am running this as administrator and checked folder security parameters, which seem ok.
Any clue what these errors are pointing to ?
Advice on how to troubleshoot?

Thank you!

**Comment by snibgo** [comment]
I assume this is a Windows BAT script.

Unfortunately, the forum software mangles the script. Please show the exact script with three back-ticks in one line above the script, and the same below the script. Or use the "Add code" button. This will ensure that special characters within the script are shown exactly as they are.

You say your script includes:
```
magick -ping "%%F" ^ -format "DoIt=%%[fx:%w==1001&&%h==1001?0:1]" ^ info:
```
But running the script shows:
```
magick -ping "Papa Roach - Last Resort.jpg"   -format "DoIt=%[fx:h==1001?0:1]
```
So Windows is not passing all the characters to ImageMagick.

You may have bugs in your use of "%".

EDIT: In addition, IM's "-fx" changed about a year ago, and it can't be used with "`-ping`".

And you seem to have stray "^" characters in your script.

**Comment by Montreal666** [comment]
yes correct BAT script.
here it is:

```
setlocal enabledelayedexpansion

for %%F in (*.jpg) do (

  for /F "usebackq" %%L in (`magick -ping "%%F" ^
    -format "DoIt=%%[fx:%w==1001&&%h==1001?0:1]" ^
    info:`) do set %%L

  if !DoIt!==1 (
    magick "%%F" -fuzz 1%% -trim +repage -resize 1001x1001^^ -gravity Center -crop 1001x1001+0+0 -format jpg -strip -gaussian-blur 0.05 -quality 80 +repage "%%F"
  )

)

```

The exact same script works perfectly on my Windows 10 pc.
however I see that the output has extra backticks before **magick** and after **info:**

_Windows 10 working:_

```
D:\Dropbox\ALBUM ART>(
for /F "usebackq" %L in (`magick -ping "Papa Roach - Last Resort.jpg"   -format "DoIt=%[fx:h==1001?0:1]"   info:`) do set %L
 if !DoIt! == 1 (magick "Papa Roach - Last Resort.jpg" -fuzz 1% -trim +repage -resize 1001x1001^ -gravity Center -crop 1001x1001+0+0 -format jpg -strip -gaussian-blur 0.05 -quality 80 +repage "Papa Roach - Last Resort.jpg" )
)
```

_Windows 11 not working:_

```
D:\Dropbox\ALBUM ART>(
for /F "usebackq" %L in (magick -ping "Papa Roach - Last Resort.jpg"   -format "DoIt=%[fx:h==1001?0:1]"   info:) do set %L
if !DoIt! == 1 (magick "Papa Roach - Last Resort.jpg" -fuzz 1% -trim +repage -resize 1001x1001^ -gravity Center -crop 1001x1001+0+0 -format jpg -strip -gaussian-blur 0.05 -quality 80 +repage "Papa Roach - Last Resort.jpg" )
)
```

So I guess understanding what the 2 errors mean would help pinpoint the issue.

magick: GetHslInt failure 0 0,0 @ error/fx.c/ExecuteRPN/3151.
magick: ExecuteRPN failed @ error/fx.c/FxEvaluateChannelExpression/4002.

Or how to restore those backticks with my new install.

Any help appreciated, 

Thx

**Comment by snibgo** [comment]
When `w` and `h` and other variables are used within `%fx[...]`, they should not be prefixed with `%`.

`%fx[...]` needs pixels, but `-ping` doesn't read pixels, giving "`GetHslInt failure`". So remove `-ping`.

So the script should now be:
```
setlocal enabledelayedexpansion

for %%F in (*.jpg) do (

  for /F "usebackq" %%L in (`magick "%%F" ^
    -format "DoIt=%%[fx:w==1001&&h==1001?0:1]" ^
    info:`) do set %%L

  if !DoIt!==1 (
    magick "%%F" -fuzz 1%% -trim +repage -resize 1001x1001^^ -gravity Center -crop 1001x1001+0+0 -format jpg -strip -gaussian-blur 0.05 -quality 80 +repage "%%F"
  )

)
```

**Comment by Montreal666** [comment]
Thank you , this works however it doesn't process subdirectories (which was the case before).
How should I modify the script to restore subdirectories processing ?

Thx for your help, much appreciated.

BTW any clue for which reason the original/unmodified script does work on another PC running Win10 ?

**Comment by snibgo** [comment]
On using `for` to traverse subdirectories, type "`help for`" at the command line. You need `/R`.

In a BAT script, when Windows finds `%%`, it should reduce that to a single `%` and pass it to the called program (here, `magick`). But when it finds a single `%` followed by another non-digit character such as `w`, it searches ahead for the next `%`, and interprets intervening characters as an environment variable. So we can write things like `%DoIt%` or `!DoIt!` and Windows will substitute the value of that variable. If the variable is not defined, it substitutes nothing, simply removing the `%...%`.

And that happened to you. You had:
```
"DoIt=%%[fx:%w==1001&&%h==1001?0:1]"
```
Windows reads `%w==1001&&%` as the name of a required environment variable `w==1001&&`. That variable is not defined, so Windows strips it. In your OP, the command was echoed back to you as:
```
"DoIt=%[fx:h==1001?0:1]"
```
As you can see, the `%w==1001&&%` was entirely removed.

**Comment by Montreal666** [comment]
Thank you, so that would be 

```
setlocal enabledelayedexpansion

for %%F in (*.jpg) do (

  for /F "usebackq" %%L in (`magick "%%F" ^
    -format "DoIt=%%[fx:w==1001&&h==1001?0:1]" ^
    info:`) do set %%L

  if !DoIt!==1 (
    magick "%%F" -fuzz 1%% -trim +repage -resize 1001x1001^^ -gravity Center -crop 1001x1001+0+0 -format jpg -strip -gaussian-blur 0.05 -quality 80 +repage "%%F" /R
  )

)
```

For the Windows/Batch file processing  I am still puzzled at why this would work on another pc running Windows10. 
Is this logic specific to Windows11?

**Comment by snibgo** [comment]
> Is this logic specific to Windows11?

No. It applies to Windows 11, and 8.1, and older versions. I have never used Windows 10.

**Reply by Montreal666** [reply]
ok that would explain it, Thanks
